components/gateway/pkg/proxy/adaptertransport.go

```diff
@@ -8,87 +8,192 @@ import (
+	ctx := context.WithValue(req.Context(), correlation.RequestIDHeaderKey, auditLogCtx.correlationHeaders)
+
+	if err := doPreAuditLog(ctx, auditLogCtx, buffer.Bytes()); nil != err {
+		return nil, errors.Wrap(err, "Failed to do pre-change auditlog")
 	}
 
+	log.C(ctx).Info("Proxying request...")
 	resp, err = t.RoundTripper.RoundTrip(req)
 	if err != nil {
 		return nil, errors.Wrap(err, "on request round trip")
 	}
+	defer httpcommon.CloseBody(req.Context(), resp.Body)
 
-	responseBody, err := ioutil.ReadAll(resp.Body)
+	responseBody, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return nil, err
 	}
 	resp.Body = ioutil.NopCloser(bytes.NewReader(responseBody))
-	defer httpcommon.CloseBody(req.Context(), resp.Body)
 
-	err = t.auditlogSink.Log(req.Context(), AuditlogMessage{
-		CorrelationIDHeaders: correlationHeaders,
-		Request:              string(requestBody),
-		Response:             string(responseBody),
-		Claims:               claims,
-	})
-	if err != nil {
-		log.C(ctx).WithError(err).Errorf("failed to send a post-change auditlog message to auditlog service: %v", err)
+	if err := doPostAuditLog(ctx, auditLogCtx, responseBody); nil != err {
+		log.C(ctx).WithError(err).Error("failed to send post-change audit-log")
 	}
 
 	return resp, nil
 }
 
+func doPreAuditLog(ctx context.Context, auditLogCtx *auditLogContext, requestBody []byte) error {
+	log.C(ctx).Infof("Maximum request body length per audit log message: %d", auditLogCtx.cfg.MsgBodySizeLimit)
+	log.C(ctx).Infof("Length of received request body: %d", len(requestBody))
+
+	transactionId := auditLogCtx.logId.String()
+
+	// Compute how many pre-audit logs we have to send
+	shards := len(requestBody) / auditLogCtx.cfg.MsgBodySizeLimit
+	if shards%auditLogCtx.cfg.MsgBodySizeLimit != 0 {
+		shards += 1
+	}
+
+	// Compute the maximum size of each auditlog message
+	shardLength := min(auditLogCtx.cfg.MsgBodySizeLimit, len(requestBody))
+	if shards > 1 {
+		shardLength = multipleOfFour(len(requestBody) / shards)
+	}
+	log.C(ctx).Infof("The request body for %s will be split to %d shards of approximately %d bytes", transactionId, shards, shardLength)
+
+	log.C(ctx).Infof("Writing pre-change auditlog for %s...", transactionId)
+	for len(requestBody) > 0 {
+		length := min(shardLength, len(requestBody))
+		shard := requestBody[:length]
+		requestBody = requestBody[length:]
+
+		err := auditLogCtx.preAuditLog.PreLog(ctx, AuditlogMessage{
+			CorrelationIDHeaders: auditLogCtx.correlationHeaders,
+			Request:              transactionId + ": " + string(shard),
+			Response:             "",
+			Claims:               auditLogCtx.claims,
+		})
+
+		if nil != err {
+			return errors.Wrapf(err, "while sending pre-change auditlog message to auditlog service for %s", transactionId)
+		}
+	}
+
+	return nil
+}
+
+func doPostAuditLog(ctx context.Context, auditLogCtx *auditLogContext, responseBody []byte) error {
```

Comment:
Looks like this function is the same as the `doPreAuditLog` func, can we extract the code into a new function? 
The only difference looks like the `PreLog` and `Log` function invocation which is in the for-loop. One idea is to write a new function that accepts a function as a parameter. And on line 94 we will pass the `PreLog` function as an argument and on line 111 we will pass the `Log` function as an argument. 
Also, this new function will not wrap the error instead, we will wrap it on the upper level.

Reply:
They are indeed very similar, but there are also a lot of small differences, which makes it very difficult to make it only one function. I'll extract some common duplicate code in a separate method